### PR TITLE
Cloud API: Add pairing functionality

### PR DESCRIPTION
**Pull request by jtguggedal**
This PR adds pairing functionality in the Cloud API.
In addition it pulls back in some functionality that was prematurely removed by me in #795 with regards to using buttons and switches to associate a device with nRF Cloud. This method is now considered legacy, and the new method (also added in this PR) will use a pin code along with the IMEI number preprogrammed in the device. However, we still have to support both.

---
net: cloud: Add association events

Cloud solutions can require to associate a particular physical
device with a cloud user account. This patch adds events for
association requests so that cloud backends can implement it
if required.

---

net: cloud: Add pointer in event structure

Some cloud events may have associated data, without being an
error or message. This patch adds a void pointer to the
data union so that backends can add custom data to an event.

---
net: lib: nrf_cloud: Add association handling with cloud API

Adds handling of association requests and responses using the
cloud API.
In addition some cleanup in event handling is done to improve
code readbility.

---

subsys: net: lib: nrf_cloud: Add DUA with PIN

Adding support for device user association with PIN code.

---

applications: asset_tracker: User association via cloud API

This patch adds cloud API support for the legacy user association
method using buttons and switches and the DK buttons and LEDs
library.
Simple socket error handling has also been added, as user
association usually will lead to an erorr on the socket and
require reboot.

### CODE REVIEW

**Comment by jtguggedal** [comment]
Marking DNM as we still have some functionality verification to do on PCA20035. Other than that, it's ready for review.

File: include/net/cloud.h [hunk 1]
```diff
@@ -78,6 +78,7 @@ struct cloud_event {
 	union {
 		struct cloud_msg msg;
 		int err;
+		void *ptr;
```
Comment (rlubos) [inline include/net/cloud.h]: Perhaps we should make this field more descriptive as we aim into a generic API? Right now, an application needs to figure out what kind of association is it dealing with, based solely on the pointer value/data. My proposal is to create something like this:
```
enum cloud_association_type {
    CLOUD_ASSOCIATION_SEQUENCE,
    CLOUD_ASSOCIATION_PIN,
}

struct cloud_association_data {
    enum cloud_association_type type;
    u8_t *buf;
    size_t len;
}
```
And then replacing the vague `void *ptr;` with `struct cloud_association_data association_data;`. IMO it would be more intuitive for the application to use. 

The naming could be a subject of change, as "association" seems longish, and shortening it to "ass" doesn't sound like a good idea :D "bond_data/bond_type" perhaps?

Comment (jtguggedal) [reply include/net/cloud.h]: Yeah, good proposal, the `*ptr` felt hackish just to get it to work (and it was, to be honest). Perhaps `pair` is an ok short name?

Comment (rlubos) [reply include/net/cloud.h]: Sounds fine for me.

Comment (jtguggedal) [reply include/net/cloud.h]: Updated per your suggestions and rebased.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud.c [hunk 2]
```diff
@@ -20,7 +20,7 @@ LOG_MODULE_REGISTER(nrf_cloud, CONFIG_NRF_CLOUD_LOG_LEVEL);
 #define NOT_VALID_STATE(EXPECTED) ((EXPECTED) < m_current_state)
 
 /* Identifier for user association message sent to the cloud.
- * A unique, unsigned 16 bit magic number. Can't be zero.
+ * A unique, unsigned 16p bit magic number. Can't be zero.
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud.c]: Was that change intended? If so what does that mean?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c [hunk 3]
```diff
@@ -388,6 +400,16 @@ int nrf_cloud_encode_state(u32_t reported_state, struct nrf_cloud_data *output)
 	ret = 0;
 
 	switch (reported_state) {
+	case STATE_UA_PIN_WAIT: {
+
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c]: Extra empty line.

File: applications/asset_tracker/src/main.c [hunk 4]
```diff
@@ -528,6 +537,99 @@ void sensors_start(void)
...
+	} else {
+		printk("LTE link disconnected\n");
+	}
+
+#ifdef CONFIG_REBOOT
+	printk("Rebooting...\n");
+	sys_reboot(SYS_REBOOT_COLD);
+#endif
+	printk("**** Manual reboot required ***\n");
+
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: Extra empty line.

File: applications/asset_tracker/src/main.c [hunk 5]
```diff
@@ -528,6 +537,99 @@ void sensors_start(void)
 	}
 }
 
+/**@brief nRF Cloud specific callback for cloud association event. */
+static void on_user_pairing_req(const struct cloud_event *evt)
+{
+	if (evt->data.pair_info.buf!= NULL) {
```
Comment (rlubos) [inline applications/asset_tracker/src/main.c]: You could check the `type` here in order to resolve whether to do the pattern matching or pin pairing. IMO that would be easier to understand.

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h [hunk 6]
```diff
@@ -37,7 +37,8 @@ int nrf_cloud_decode_requested_state(const struct nrf_cloud_data *payload,
 /**@brief Decodes data endpoint information. */
 int nrf_cloud_decode_data_endpoint(const struct nrf_cloud_data *input,
 				   struct nrf_cloud_data *tx_endpoint,
-				   struct nrf_cloud_data *rx_endpoint);
+				   struct nrf_cloud_data *rx_endpoint,
+				   struct nrf_cloud_data *endpoint);
```
Comment (rlubos) [inline subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h]: So this new `endpoint` - is it only related to PIN association or is it something more? Perhaps we could use a more precise name for it? `endpoint` is pretty ambiguous and hard to follow in the code below.

Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h]: It's supposed to replace both rx and tx endpoints in the future, so there will be only one endpoint. For now, this is only future-proofing, as more changes are required both on the cloud and device end of things.

Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h]: I'll have another look at the naming

File: subsys/net/lib/nrf_cloud/src/nrf_cloud.c (hunk 2 above)
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud.c]: Definitely not intended. Will fix.

File: applications/asset_tracker/src/main.c (hunk 5 above)
Comment (jtguggedal) [reply applications/asset_tracker/src/main.c]: Yeah, forgot to update that when the new struct was added. Will fix, thanks

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c (hunk 3 above)
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c]: Fixed

File: subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h (hunk 6 above)
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h]: Called it `m_endpoint`, as that's what it's called on the AWS side of things

Comment (rlubos) [reply subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h]: Looks dangerously like Hungarian notation, but if it's mirroring the AWS I think it's fine.

File: applications/asset_tracker/src/main.c [hunk 7]
```diff
@@ -164,7 +173,7 @@ void error_handler(enum error_type err_type, int err_code)
 #endif
 	}
 
-#if !defined(CONFIG_DEBUG)
+#if !defined(CONFIG_DEBUG) && defined(CONFIG_REBOOT)
 	sys_reboot(SYS_REBOOT_COLD);
```
Comment (metc) [inline applications/asset_tracker/src/main.c]: Do we need a `LOG_PANIC();` here?

Comment (jtguggedal) [reply applications/asset_tracker/src/main.c]: Good idea, added.

**Comment by rlubos** [comment]
@metc Could you re-review, any other comments?
